src/NodeServices/Interfaces.cs

```diff
@@ -188,4 +191,108 @@ internal interface IExtensionApplication
         /// <param name="baseDirectory"></param>
         void PreloadAsmLibraries(string baseDirectory);
     }
+
+    //TODO should this live in DynamoServices?
+    //make these internal somehow?
+    public interface IExternalCodeCompletionProviderCore
+    {
+
+        /// <summary>
+        /// Already discovered variable types
+        /// </summary>
+        Dictionary<string, Type> VariableTypes { get; set; }
+
+        /// <summary>
+        /// Types that have already been imported into the scope
+        /// </summary>
+        Dictionary<string, Type> ImportedTypes { get; set; }
+
+        /// <summary>
+        /// Generate completion data for the specified text, while import the given types into the
+        /// scope and discovering variable assignments.
+        /// </summary>
+        /// <param name="code">The code to parse</param>
+        /// <param name="expand">Determines if the entire namespace should be used</param>
+        /// <returns>Return a list of IronPythonCompletionData </returns>
+        IExternalCodeCompletionData[] GetCompletionData(string code, bool expand = false);
+
+        /// <summary>
+        /// Try to generate a description from a typename
+        /// </summary>
+        /// <param name="stub">Everything before the last namespace or type name e.g. System.Collections in System.Collections.ArrayList</param>
+        /// <param name="item">Everything after the stub</param>
+        /// <param name="isInstance">Whether it's an instance or not</param>
+        string GetDescription(string stub, string item, bool isInstance);
+
+        bool MatchingEngine(string engineName);
+
+        //TODO think about this, more general?
+        void ImportStdLibrary(string dynamoCorePath);
+
+        void Log(string message);
+
+    }
+
+
+    /// <summary>
+    /// This interface is essentially the same as ICompletionData from AvalonEdit 
+    /// except it does not include any references to WPF.
+    /// </summary>
+    public interface IExternalCodeCompletionData
+    {
+        //TODO hmmmmmm what about image path or image instead?
+      
+        ExternalCodeCompletionType CompletionType  { get; }
+        //
+        // Summary:
+        //     Gets the text. This property is used to filter the list of visible elements.
+        string Text { get; }
+        //
+        // Summary:
+        //     The displayed content. This can be the same as 'Text', or a WPF UIElement if
+        //     you want to display rich content.
+        object Content { get; }
+        //
+        // Summary:
+        //     Gets the description.
+        object Description { get; }
```

Comment:
Can this also be a WPF UIElement?

Reply:
I think in our case it's always text, but the `ICompletionData` interface from Avalon has this typed as an object, and I was following that pattern, I'll see if we can make it a string.

Reply:
it compiles with this changed to string - I will do some testing while I am adding additional tests for the new classes.